### PR TITLE
Implement address sweeper

### PR DESCRIPTION
**Pull request by stdavis**
This is very much a work-in-progress that I'm looking for feedback on. I'm still trying to figure out what direction this sweeper should go. There's a lot of differences between it and the existing ones. Here are a few questions for me to figure out:

- [x] For example, what should be marked as issues? Missing parts of the address (#28) only? Non-standardized cardinal directions, street types and P.O. Boxes?
- [x] Try-Fix: what, exactly, is the use-case for this? Should it normalize all addresses or just those that have missing parts? If it's the latter, what good is it? Normalized bad addresses will still be bad.
- [x] This report should give some sort of report on what data was changed/removed rather than just a list of the errors that happened during the try-fix.

This PR also takes input from #36. Specifically, it removes the class property `report` from the sweeper class and builds a new one for each `sweep()` and `try_fix()`. This is just one possible way to solve the issue. I'm not yet sure that it's the best way.

- [x] If the above changes to the report dictionary are ratified, I'll need to tweak the existing sweepers to match it.

With all this being said, the address sweeper is functional. For example:
```
(sweeper) λ python -m sweeper sweep addresses --workspace=C:\temp\addresses.gdb --table-name=Normalize --try-fix --field-name=StreetAddress
running 1 sweepers. try fix: True
Address Test Report for Normalize
2 issues found
  Issues found:
    ObjectID 3: "123" is not recognized as a valid street address, or P.O. Box
    ObjectID 5: missing address parts: street_name

Select statement to view issues in ArcGIS:
OBJECTID IN (3, 5)
---
Address Try Fix Report for Normalize
2 issues found
  Issues found:
    ObjectID 3: "123" is not recognized as a valid street address, or P.O. Box
    ObjectID 5: 123 south -> 123 S

Select statement to view issues in ArcGIS:
OBJECTID IN (3, 5)
---
Address Test Report for Normalize
2 issues found
  Issues found:
    ObjectID 3: "123" is not recognized as a valid street address, or P.O. Box
    ObjectID 5: missing address parts: street_name

Select statement to view issues in ArcGIS:
OBJECTID IN (3, 5)
---
```

Finally, there are a few niceties sprinkled into this PR such as DRY-ing up the version number and easier-to-read formatting of the docopt text.

### CODE REVIEW

**Comment by steveoh** [comment]
I vote that records should be marked as issues when there are missing parts of the address.
I vote that try fix is not implemented for this sweeper. 
I like the report output that you are generating.

**Comment by stdavis** [comment]
If you look at this from the use-case outside of geocoding, just normalizing addresses, then it could be useful for things like standardizing cardinal directions and street types. Maybe that's all that try_fix should worry about. I'm not sure if that would be useful to anyone or not.

**Comment by steveoh** [comment]
I edited my answer

**Comment by stdavis** [comment]
> I vote that records should be marked as issues when there are missing parts of the address **AND** using aggregated data from the road grinder project, provide warnings/reasonings why an address may not geocode or may lose points.

I think that this is a good idea but I'm not sure that it is worth the effort without some clear direction from users. At the very least I would propose that we punt it to a future issue.

I did think that try_fix might be useful to users for normalizing addresses and it was an easy implementation so I did it. With the latest changes if you feed it this table:

image

the tool will update the values to be:

image

and generate this report:
```
(sweeper) λ python -m sweeper sweep addresses --workspace=C:\temp\addresses.gdb --table-name=Normalize --try-fix --field-name=StreetAddress
running 1 sweepers. try fix: True
Address Test Report for Normalize
5 issues found:
    ObjectID 1: address not fully normalized: "123 Main" -> "123 MAIN"
    ObjectID 2: address not fully normalized: "123 So. Main Street" -> "123 S MAIN ST"
    ObjectID 3: "123" is not recognized as a valid street address, or P.O. Box
    ObjectID 4: address not fully normalized: "P.O. Box 321" -> "PO BOX 321"
    ObjectID 5: missing address parts: street_name

Select statement to view issues in ArcGIS:
OBJECTID IN (1, 2, 3, 4, 5)
---
Address Try Fix Report for Normalize
4 fixes found:
    ObjectID 1: 123 Main -> 123 MAIN
    ObjectID 2: 123 So. Main Street -> 123 S MAIN ST
    ObjectID 4: P.O. Box 321 -> PO BOX 321
    ObjectID 5: 123 south -> 123 S

Select statement to view fixes in ArcGIS:
OBJECTID IN (1, 2, 4, 5)
---
1 issues found:
    ObjectID 3: "123" is not recognized as a valid street address, or P.O. Box

Select statement to view issues in ArcGIS:
OBJECTID IN (3)
---
Address Test Report for Normalize
2 issues found:
    ObjectID 3: "123" is not recognized as a valid street address, or P.O. Box
    ObjectID 5: missing address parts: street_name

Select statement to view issues in ArcGIS:
OBJECTID IN (3, 5)
---
```

I updated the other sweepers to use the new report format so I think that this PR is ready to merge and then we can build upon it later.

**Comment by stdavis** [comment]
@steveoh I'd love for a review on this when you get a chance.

File: src/sweeper/__main__.py [hunk 1]
```diff
@@ -4,34 +4,39 @@
...
 from .sweepers.duplicates import DuplicateTest
 from .sweepers.empties import EmptyTest
+from .sweepers.addresses import AddressTest
 
 
 def main():
     '''Main entry point for program. Parse arguments and pass to sweeper modules.
     '''
-    args = docopt(__doc__, version='1.0.3')
+    args = docopt(__doc__, version=pkg_resources.require('agrc-sweeper')[0].version)
```
Comment (steveoh) [inline src/sweeper/__main__.py]: 👍

File: src/sweeper/sweepers/addresses.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                        parsed_address = Address(address)
+                    except Exception as exception:
+                        issue_message = str(exception)
+
+                    if issue_message is None:
+                        parts_found = set(parsed_address.__dict__)
+                        missing_parts = required_street_address_parts - parts_found
+                        if len(missing_parts) > 0 and 'po_box' not in parts_found:
+                            issue_message = f'missing address parts: {", ".join(missing_parts)}'
+                        elif parsed_address.normalized != address:
```
Comment (steveoh) [inline src/sweeper/sweepers/addresses.py]: can you explain this part to me?

**Pull request merged**